src/gt4py/next/embedded/nd_array_field.py

```diff
@@ -203,15 +203,13 @@ def remap(
 
     __call__ = remap  # type: ignore[assignment]
 
-    def restrict(self, index: common.AnyIndexSpec) -> common.Field | core_defs.ScalarT:
+    def restrict(self, index: common.AnyIndexSpec) -> common.Field:
```

Comment:
The return type annotation should also be changed in all the related protocols like `Field`, `MutableField`, ...

Reply:
`ConstantField` returns a scalar for `restrict` and `_tuple_assign_field` uses `MutableField` to return scalars. I would leave these as they are

Reply:
They should be changed then. In the end if you slice any field, you want to get a field back.